src/control/StateMachine.ts

```diff
@@ -0,0 +1,82 @@
+/**
+ * Copyright (c) 2015 NAVER Corp.
+ * egjs projects are licensed under the MIT license
+ */
+
+import { AxesEvents } from "@egjs/axes";
+
+import IdleState from "./states/IdleState";
+import HoldingState from "./states/HoldingState";
+import DraggingState from "./states/DraggingState";
+import AnimatingState from "./states/AnimatingState";
+import DisabledState from "./states/DisabledState";
+
+import Flicking from "~/Flicking";
+import State from "~/control/states/State";
+import { AXES_EVENT, STATE_TYPE } from "~/const/internal";
+
+class StateMachine {
+  private _state: State;
+
+  public constructor() {
+    this._state = new IdleState();
+  }
+
+  public getState(): State {
+    return this._state;
+  }
+
+  public fire(eventType: keyof AxesEvents, externalCtx: {
+    flicking: Flicking;
+    axesEvent: any;
+  }) {
+    const currentState = this._state;
+    const ctx = { ...externalCtx, transitTo: this.transitTo };
+
+    switch (eventType) {
+      case AXES_EVENT.HOLD:
+        currentState.onHold(ctx);
+        break;
+      case AXES_EVENT.CHANGE:
+        currentState.onChange(ctx);
+        break;
+      case AXES_EVENT.RELEASE:
+        currentState.onRelease(ctx);
+        break;
+      case AXES_EVENT.ANIMATION_END:
+        currentState.onAnimationEnd(ctx);
+        break;
+      case AXES_EVENT.FINISH:
+        currentState.onFinish(ctx);
+        break;
```

Comment:
Well if it's not a file size problem, I personally prefer that style as I think it can explain the meaning of the code better.
It is also a more typescript-friendly approach.

Reply:
In general, in the case of switch statements, when the type increases, the probability of generating an error is high, so it is better to improve. However, in the current case, it doesn't seem to be bad even now, since it actually converts the old code. You may find it helpful later by referring to the article below.

- https://stackoverflow.com/questions/54079960/refactoring-switch-statement-typescript
- https://medium.com/itchis-coffee-shop/how-to-refactor-a-switch-statement-with-object-composition-in-angular-ba2bc308b7d2

Reply:
I don't agree with that. Not all switch statements are bad.
The second link that you gave tells that switch statements are bad for the following two reasons.

### 1. It's error-prone, because related codes should be combined to each case.

The author of that article insists that **Object Composition & Factory Method** is the solution to bypass switch. 
But as you can see I'm using the exact same thing in my code. `transitTo` is the factory method itself, and I'm using object composition in each state(yes, this is also a state pattern).
And for the `fire`, it just propagates the Axes event to the state. This is also a case of object composition by using state.
I said my code is more typescript-friendly because changing code like that will make the state call `any` typed method.
![image](https://user-images.githubusercontent.com/26213435/105839629-2538ef00-6015-11eb-8f95-2c241592b790.png)
And also, calling methods like that is more error-prone as just one typo can create all kinds of errors, and it can't be statically caught.

### 2. It's time-comsuming, as adding more cases will take some time
I'm half-agree with this. I agree that adding more cases will take more time, but in this case, we're selectively choosing Axes event to listen to, which is very limited and also have to be precise.
And as we're using Typescript, we can easily notice the missing cases. Like, I'm not using the `default` case in my code but still have no ts/lint error because I'm handling all the cases given. Adding more types will emerge the static error, so we can easily handle that.
![image](https://user-images.githubusercontent.com/26213435/105840510-7ac1cb80-6016-11eb-9226-6e63673aa43c.png)
> Simply, adding more enum value shows error on my IDE

Reply:
I agree on some parts. So first of all, I don't want to say that I have to do that. The point I want to say is that in the case of a switch statement, the points that change according to the added state increase. For example, when the type of state.SUPER increases in the swtich statement, the point that needs to be modified is the part of adding the switch statement and the corresponding method. If the switch is changed to the example above, it is said that it is good because the point to be concerned with is reduced from 2 to 1 because only objects that match the state are created. This is because there is a high probability of generating errors if the number of parts that are changed according to some change in development increases.

However, the reason I didn't think this part was very important is that it is already stable code, so the state change is unlikely, so it is good even now. :)